File: sdk/desktop/macos/Tests/KsealDesktopTests/EnterprisePolicyTests.swift
import XCTest
@testable import KsealDesktop

final class EnterprisePolicyTests: XCTestCase {

    func testStrictBaselineRelaxesNothing() {
        let p = EnterprisePolicy.strict
        XCTAssertTrue(p.isStrict)
        XCTAssertFalse(p.permitDebugger)
        XCTAssertTrue(p.injectionAllowlist.isEmpty)
        XCTAssertEqual(p.telemetryVerbosity, .standard)
        XCTAssertFalse(p.requireHardwareBackedProofKey)
        XCTAssertFalse(p.allowsModule("/anything"))
    }

    func testPartialJsonKeepsStrictDefaultsForMissingKeys() throws {
        let json = Data(#"{ "permitDebugger": true }"#.utf8)
        let p = try JSONDecoder().decode(EnterprisePolicy.self, from: json)
        XCTAssertTrue(p.permitDebugger)
        // Unspecified controls stay strict.
        XCTAssertTrue(p.injectionAllowlist.isEmpty)
        XCTAssertEqual(p.telemetryVerbosity, .standard)
        XCTAssertFalse(p.requireHardwareBackedProofKey)
        XCTAssertFalse(p.isStrict)
    }

    func testFullJsonRoundTrip() throws {
        let json = Data(#"""
        {
          "permitDebugger": true,
          "injectionAllowlist": ["/Library/Acme/plugin.dylib", "/opt/agents/"],
          "telemetryVerbosity": "minimal",
          "requireHardwareBackedProofKey": true
        }
        """#.utf8)
        let p = try JSONDecoder().decode(EnterprisePolicy.self, from: json)
        XCTAssertTrue(p.permitDebugger)
        XCTAssertEqual(p.telemetryVerbosity, .minimal)
        XCTAssertTrue(p.requireHardwareBackedProofKey)
        XCTAssertEqual(p.injectionAllowlist.count, 2)
    }

    func testAllowsModuleExactAndPrefix() {
        let p = EnterprisePolicy(injectionAllowlist: ["/Library/Acme/plugin.dylib", "/opt/agents/"])
        XCTAssertTrue(p.allowsModule("/Library/Acme/plugin.dylib"))      // exact
        XCTAssertTrue(p.allowsModule("/opt/agents/telemetry.dylib"))     // prefix
        XCTAssertFalse(p.allowsModule("/opt/agents"))                    // prefix needs trailing /
        XCTAssertFalse(p.allowsModule("/tmp/evil.dylib"))                // not listed
        XCTAssertFalse(p.allowsModule("/Library/Acme/plugin.dylib.bak")) // exact only
    }

    func testFileProviderReadsPolicy() throws {
        let url = FileManager.default.temporaryDirectory
            .appendingPathComponent("kseal-policy-\(UUID().uuidString).json")
        try Data(#"{ "telemetryVerbosity": "verbose" }"#.utf8).write(to: url)
        let provider = FileEnterprisePolicyProvider(url: url)
        XCTAssertEqual(provider.currentPolicy().telemetryVerbosity, .verbose)
    }

    func testFileProviderFailsSafeOnMissingFile() {
        let url = URL(fileURLWithPath: "/nonexistent/\(UUID().uuidString).json")
        XCTAssertTrue(FileEnterprisePolicyProvider(url: url).currentPolicy().isStrict)
    }

    func testFileProviderFailsSafeOnMalformedJson() throws {
        let url = FileManager.default.temporaryDirectory
            .appendingPathComponent("kseal-bad-\(UUID().uuidString).json")
        try Data("not json".utf8).write(to: url)
        XCTAssertTrue(FileEnterprisePolicyProvider(url: url).currentPolicy().isStrict)
    }

    func testStaticProvider() {
        let p = EnterprisePolicy(permitDebugger: true)
        XCTAssertEqual(StaticEnterprisePolicyProvider(p).currentPolicy(), p)
        XCTAssertTrue(StaticEnterprisePolicyProvider().currentPolicy().isStrict)
    }
}

File: sdk/desktop/windows/tests/Kseal.Desktop.Tests/EnterprisePolicyTests.cs
using System.Text;
using Xunit;

namespace Kseal.Desktop.Tests;

public class EnterprisePolicyTests : IDisposable
{
    private readonly string _dir = Path.Combine(Path.GetTempPath(), $"kseal-pol-{Guid.NewGuid():N}");

    public EnterprisePolicyTests() => Directory.CreateDirectory(_dir);

    public void Dispose()
    {
        try { Directory.Delete(_dir, recursive: true); } catch (IOException) { }
        GC.SuppressFinalize(this);
    }

    [Fact]
    public void StrictBaselineRelaxesNothing()
    {
        var p = EnterprisePolicy.Strict;
        Assert.True(p.IsStrict);
        Assert.False(p.PermitDebugger);
        Assert.Empty(p.InjectionAllowlist);
        Assert.Equal(TelemetryVerbosity.Standard, p.TelemetryVerbosity);
        Assert.False(p.RequireHardwareBackedProofKey);
        Assert.False(p.AllowsModule(@"C:\anything.dll"));
    }

    [Fact]
    public void PartialJsonKeepsStrictDefaultsForMissingKeys()
    {
        var p = EnterprisePolicy.FromJson("""{ "permitDebugger": true }"""u8.ToArray());
        Assert.True(p.PermitDebugger);
        Assert.Empty(p.InjectionAllowlist);
        Assert.Equal(TelemetryVerbosity.Standard, p.TelemetryVerbosity);
        Assert.False(p.RequireHardwareBackedProofKey);
        Assert.False(p.IsStrict);
    }

    [Fact]
    public void FullJsonRoundTrip()
    {
        byte[] json = Encoding.UTF8.GetBytes("""
        {
          "permitDebugger": true,
          "injectionAllowlist": ["C:\\Program Files\\Acme\\plugin.dll", "C:\\agents\\"],
          "telemetryVerbosity": "minimal",
          "requireHardwareBackedProofKey": true
        }
        """);
        var p = EnterprisePolicy.FromJson(json);
        Assert.True(p.PermitDebugger);
        Assert.Equal(TelemetryVerbosity.Minimal, p.TelemetryVerbosity);
        Assert.True(p.RequireHardwareBackedProofKey);
        Assert.Equal(2, p.InjectionAllowlist.Count);
    }

    [Fact]
    public void MalformedJsonFailsSafeToStrict()
    {
        Assert.True(EnterprisePolicy.FromJson("not json"u8.ToArray()).IsStrict);
    }

    [Fact]
    public void AllowsModuleExactAndPrefixForBothSeparators()
    {
        var p = new EnterprisePolicy
        {
            InjectionAllowlist = [@"C:\Program Files\Acme\plugin.dll", @"C:\agents\", "/opt/agents/"],
        };
        Assert.True(p.AllowsModule(@"C:\Program Files\Acme\plugin.dll"));   // exact
        Assert.True(p.AllowsModule(@"C:\agents\telemetry.dll"));           // backslash prefix
        Assert.True(p.AllowsModule("/opt/agents/telemetry.so"));          // forward-slash prefix
        Assert.False(p.AllowsModule(@"C:\agents"));                        // prefix needs trailing sep
        Assert.False(p.AllowsModule(@"C:\Windows\evil.dll"));              // not listed
    }

    [Fact]
    public void AllowsModuleMatchesCasePerPlatform()
    {
        var p = new EnterprisePolicy
        {
            InjectionAllowlist = [@"C:\Program Files\Acme\plugin.dll", @"C:\agents\"],
        };
        // Windows module paths are case-insensitive; other hosts are case-sensitive.
        bool expected = OperatingSystem.IsWindows();
        Assert.Equal(expected, p.AllowsModule(@"c:\program files\acme\plugin.dll"));
        Assert.Equal(expected, p.AllowsModule(@"C:\AGENTS\telemetry.dll"));
    }

    [Fact]
    public void AllowsModuleRejectsParentTraversal()
    {
        var p = new EnterprisePolicy { InjectionAllowlist = [@"C:\agents\"] };
        // A '..' segment that escapes the allowlisted prefix must not be allowlisted.
        Assert.False(p.AllowsModule(@"C:\agents\..\evil.dll"));
        Assert.False(p.AllowsModule(@"C:\agents\sub\..\..\evil.dll"));
    }

    [Fact]
    public void FileProviderReadsPolicy()
    {
        string path = Path.Combine(_dir, "policy.json");
        File.WriteAllText(path, """{ "telemetryVerbosity": "verbose" }""");
        Assert.Equal(TelemetryVerbosity.Verbose, new FileEnterprisePolicyProvider(path).CurrentPolicy().TelemetryVerbosity);
    }

    [Fact]
    public void FileProviderFailsSafeOnMissingFile()
    {
        var provider = new FileEnterprisePolicyProvider(Path.Combine(_dir, "missing.json"));
        Assert.True(provider.CurrentPolicy().IsStrict);
    }

    [Fact]
    public void FileProviderFailsSafeOnMalformedFile()
    {
        string path = Path.Combine(_dir, "bad.json");
        File.WriteAllText(path, "not json");
        Assert.True(new FileEnterprisePolicyProvider(path).CurrentPolicy().IsStrict);
    }

    [Fact]
    public void StaticProvider()
    {
        var p = new EnterprisePolicy { PermitDebugger = true };
        Assert.Equal(p, new StaticEnterprisePolicyProvider(p).CurrentPolicy());
        Assert.True(new StaticEnterprisePolicyProvider().CurrentPolicy().IsStrict);
    }
}
